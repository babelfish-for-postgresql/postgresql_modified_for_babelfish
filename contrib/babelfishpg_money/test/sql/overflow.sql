-- Ensure the expected extreme values can be represented
SELECT '-92233720368547758.08'::FIXEDDECIMAL as minvalue,'92233720368547758.07'::FIXEDDECIMAL as maxvalue;

SELECT '-92233720368547758.09'::FIXEDDECIMAL;

SELECT '92233720368547758.08'::FIXEDDECIMAL;

-- Ensure casts from numeric to fixeddecimal work
SELECT '92233720368547758.07'::numeric::FIXEDDECIMAL;

-- The literal below must be quoted as the parser seems to read the literal as
-- a positive number first and then us the - unary operator to make it negaive.
-- This would overflow without the quotes as this number cannot be represented
-- in a positive fixeddecimal.
SELECT '-92233720368547758.08'::numeric::FIXEDDECIMAL;

-- Ensure casts from numeric to fixed decimal detect overflow
SELECT '92233720368547758.08'::numeric::FIXEDDECIMAL;

SELECT '-92233720368547758.09'::numeric::FIXEDDECIMAL;

SELECT '-92233720368547758.08'::FIXEDDECIMAL - '0.01'::FIXEDDECIMAL;

SELECT '92233720368547758.07'::FIXEDDECIMAL + '0.01'::FIXEDDECIMAL;

-- Should not overflow
SELECT '46116860184273879.03'::FIXEDDECIMAL * '2.00'::FIXEDDECIMAL;

-- Ensure this overflows
SELECT '46116860184273879.04'::FIXEDDECIMAL * '2.00'::FIXEDDECIMAL;

-- Should not overflow
SELECT '46116860184273879.03'::FIXEDDECIMAL / '0.50'::FIXEDDECIMAL;

-- Ensure this overflows
SELECT '46116860184273879.04'::FIXEDDECIMAL / '0.50'::FIXEDDECIMAL;

-- Ensure limits of int2 can be represented
SELECT '32767'::FIXEDDECIMAL::INT2,'-32768'::FIXEDDECIMAL::INT2;

-- Ensure overflow of int2 is detected
SELECT '32768'::FIXEDDECIMAL::INT2;

-- Ensure underflow of int2 is detected
SELECT '-32769'::FIXEDDECIMAL::INT2;

-- Ensure limits of int4 can be represented
SELECT '2147483647'::FIXEDDECIMAL::INT4,'-2147483648'::FIXEDDECIMAL::INT4;

-- Ensure overflow of int4 is detected
SELECT '2147483648'::FIXEDDECIMAL::INT4;

-- Ensure underflow of int4 is detected
SELECT '-2147483649'::FIXEDDECIMAL::INT4;

-- Ensure overflow is detected
SELECT SUM(a) FROM (VALUES('92233720368547758.07'::FIXEDDECIMAL),('0.01'::FIXEDDECIMAL)) a(a);

-- Ensure underflow is detected
SELECT SUM(a) FROM (VALUES('-92233720368547758.08'::FIXEDDECIMAL),('-0.01'::FIXEDDECIMAL)) a(a);

-- Test typmods
SELECT 12345.33::FIXEDDECIMAL(3,2); -- Fail

SELECT 12345.33::FIXEDDECIMAL(5,2); -- Fail

-- scale of 2 should be enforced.
SELECT 12345.44::FIXEDDECIMAL(7,0);

-- should work.
SELECT 12345.33::FIXEDDECIMAL(7,2);

-- error, precision limit should be 17
SELECT 12345.33::FIXEDDECIMAL(18,2);

CREATE TABLE fixdec (d FIXEDDECIMAL(3,2));
INSERT INTO fixdec VALUES(12.34); -- Fail
INSERT INTO fixdec VALUES(1.23); -- Pass
DROP TABLE fixdec;
