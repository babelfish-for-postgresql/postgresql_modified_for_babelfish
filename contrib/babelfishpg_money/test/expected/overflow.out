-- Ensure the expected extreme values can be represented
SELECT '-92233720368547758.08'::FIXEDDECIMAL as minvalue,'92233720368547758.07'::FIXEDDECIMAL as maxvalue;
       minvalue        |       maxvalue       
-----------------------+----------------------
 -92233720368547758.08 | 92233720368547758.07
(1 row)

SELECT '-92233720368547758.09'::FIXEDDECIMAL;
ERROR:  value "-92233720368547758.09" is out of range for type fixeddecimal
LINE 1: SELECT '-92233720368547758.09'::FIXEDDECIMAL;
               ^
SELECT '92233720368547758.08'::FIXEDDECIMAL;
ERROR:  value "92233720368547758.08" is out of range for type fixeddecimal
LINE 1: SELECT '92233720368547758.08'::FIXEDDECIMAL;
               ^
-- Ensure casts from numeric to fixeddecimal work
SELECT '92233720368547758.07'::numeric::FIXEDDECIMAL;
     fixeddecimal     
----------------------
 92233720368547758.07
(1 row)

-- The literal below must be quoted as the parser seems to read the literal as
-- a positive number first and then us the - unary operator to make it negaive.
-- This would overflow without the quotes as this number cannot be represented
-- in a positive fixeddecimal.
SELECT '-92233720368547758.08'::numeric::FIXEDDECIMAL;
     fixeddecimal      
-----------------------
 -92233720368547758.08
(1 row)

-- Ensure casts from numeric to fixed decimal detect overflow
SELECT '92233720368547758.08'::numeric::FIXEDDECIMAL;
ERROR:  value "92233720368547758.08" is out of range for type fixeddecimal
SELECT '-92233720368547758.09'::numeric::FIXEDDECIMAL;
ERROR:  value "-92233720368547758.09" is out of range for type fixeddecimal
SELECT '-92233720368547758.08'::FIXEDDECIMAL - '0.01'::FIXEDDECIMAL;
ERROR:  fixeddecimal out of range
SELECT '92233720368547758.07'::FIXEDDECIMAL + '0.01'::FIXEDDECIMAL;
ERROR:  fixeddecimal out of range
-- Should not overflow
SELECT '46116860184273879.03'::FIXEDDECIMAL * '2.00'::FIXEDDECIMAL;
       ?column?       
----------------------
 92233720368547758.06
(1 row)

-- Ensure this overflows
SELECT '46116860184273879.04'::FIXEDDECIMAL * '2.00'::FIXEDDECIMAL;
ERROR:  fixeddecimal out of range
-- Should not overflow
SELECT '46116860184273879.03'::FIXEDDECIMAL / '0.50'::FIXEDDECIMAL;
       ?column?       
----------------------
 92233720368547758.06
(1 row)

-- Ensure this overflows
SELECT '46116860184273879.04'::FIXEDDECIMAL / '0.50'::FIXEDDECIMAL;
ERROR:  fixeddecimal out of range
-- Ensure limits of int2 can be represented
SELECT '32767'::FIXEDDECIMAL::INT2,'-32768'::FIXEDDECIMAL::INT2;
 int2  |  int2  
-------+--------
 32767 | -32768
(1 row)

-- Ensure overflow of int2 is detected
SELECT '32768'::FIXEDDECIMAL::INT2;
ERROR:  smallint out of range
-- Ensure underflow of int2 is detected
SELECT '-32769'::FIXEDDECIMAL::INT2;
ERROR:  smallint out of range
-- Ensure limits of int4 can be represented
SELECT '2147483647'::FIXEDDECIMAL::INT4,'-2147483648'::FIXEDDECIMAL::INT4;
    int4    |    int4     
------------+-------------
 2147483647 | -2147483648
(1 row)

-- Ensure overflow of int4 is detected
SELECT '2147483648'::FIXEDDECIMAL::INT4;
ERROR:  integer out of range
-- Ensure underflow of int4 is detected
SELECT '-2147483649'::FIXEDDECIMAL::INT4;
ERROR:  integer out of range
-- Ensure overflow is detected
SELECT SUM(a) FROM (VALUES('92233720368547758.07'::FIXEDDECIMAL),('0.01'::FIXEDDECIMAL)) a(a);
ERROR:  fixeddecimal out of range
-- Ensure underflow is detected
SELECT SUM(a) FROM (VALUES('-92233720368547758.08'::FIXEDDECIMAL),('-0.01'::FIXEDDECIMAL)) a(a);
ERROR:  fixeddecimal out of range
-- Test typmods
SELECT 12345.33::FIXEDDECIMAL(3,2); -- Fail
ERROR:  FIXEDDECIMAL field overflow
DETAIL:  A field with precision 5, scale 2 must round to an absolute value less than 10^1.
SELECT 12345.33::FIXEDDECIMAL(5,2); -- Fail
ERROR:  FIXEDDECIMAL field overflow
DETAIL:  A field with precision 5, scale 2 must round to an absolute value less than 10^3.
-- scale of 2 should be enforced.
SELECT 12345.44::FIXEDDECIMAL(7,0);
ERROR:  FIXEDDECIMAL scale must be 2
LINE 1: SELECT 12345.44::FIXEDDECIMAL(7,0);
                         ^
-- should work.
SELECT 12345.33::FIXEDDECIMAL(7,2);
 fixeddecimal 
--------------
 12345.33
(1 row)

-- error, precision limit should be 17
SELECT 12345.33::FIXEDDECIMAL(18,2);
ERROR:  FIXEDDECIMAL precision 18 must be between 2 and 17
LINE 1: SELECT 12345.33::FIXEDDECIMAL(18,2);
                         ^
CREATE TABLE fixdec (d FIXEDDECIMAL(3,2));
INSERT INTO fixdec VALUES(12.34); -- Fail
ERROR:  FIXEDDECIMAL field overflow
DETAIL:  A field with precision 2, scale 2 must round to an absolute value less than 10^1.
INSERT INTO fixdec VALUES(1.23); -- Pass
DROP TABLE fixdec;
